dev-tools/oxlog/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+        for uuid in &pools.external {
+            let zones_path: Utf8PathBuf =
+                ["/pool/ext", &uuid.to_string(), "crypt/zone"].iter().collect();
+            // Find the zones on the given pool
+            let Ok(entries) = read_dir(zones_path.as_path()) else {
+                continue;
+            };
+            for entry in entries {
+                let Ok(zone_entry) = entry else {
+                    continue;
+                };
+                let zone = zone_entry.file_name();
+                let Some(zone) = zone.to_str() else {
+                    // not utf8
+                    continue;
+                };
+                // Load the current logs
+                let mut dir = zones_path.clone();
+                dir.push(zone);
+                dir.push("root/var/svc/log");
+                let mut paths =
+                    Paths { primary: dir, debug: vec![], extra: vec![] };
+
+                // Load the extra logs
+                if zone.starts_with("oxz_cockroachdb") {
+                    let mut dir = zones_path.clone();
+                    dir.push(zone);
+                    dir.push("root/data/logs");
+                    paths.extra.push(("cockroachdb", dir));
+                }
+
+                zones.insert(zone.to_string(), paths);
+            }
+        }
+
+        // Load the debug logs
+        for uuid in &pools.external {
+            let zones_path: Utf8PathBuf =
+                ["/pool/ext", &uuid.to_string(), "crypt/debug"]
+                    .iter()
+                    .collect();
+            // Find the zones on the given pool
+            let Ok(entries) = read_dir(zones_path.as_path()) else {
+                continue;
+            };
+            for entry in entries {
+                let Ok(zone_entry) = entry else {
+                    continue;
+                };
+                let zone = zone_entry.file_name();
+                let Some(zone) = zone.to_str() else {
+                    // not utf8
+                    continue;
+                };
+                let mut dir = zones_path.clone();
+                dir.push(zone);
+
+                // We only add debug paths if the zones have primary paths
+                zones.get_mut(zone).map(|paths| paths.debug.push(dir));
+            }
+        }
+
+        Ok(Zones { zones })
+    }
+
+    pub fn zone_logs(&self, zone: &str) -> BTreeMap<ServiceName, SvcLogs> {
+        let mut output = BTreeMap::new();
+        let Some(paths) = self.zones.get(zone) else {
+            return BTreeMap::new();
+        };
+        load_svc_logs(paths.primary.clone(), &mut output);
+        for dir in paths.debug.clone() {
+            load_svc_logs(dir, &mut output);
+        }
+        for (svc_name, dir) in paths.extra.clone() {
+            load_extra_logs(dir, svc_name, &mut output);
+        }
+        output
+    }
+}
+
+fn load_svc_logs(dir: Utf8PathBuf, logs: &mut BTreeMap<ServiceName, SvcLogs>) {
```

Comment:
There is some similar logic [here](https://github.com/oxidecomputer/omicron/blob/86c64a2fe8167de354a827425b2d48cfe0f33f33/illumos-utils/src/running_zone.rs#L1419) for finding logs during zone-bundling, it might be a good idea to share it somehow. Calling into this library in the `running_zone` mod seems reasonable.

Reply:
I can certainly add a helper function to do this. However, right now oxlog only looks at logs with the `oxide-` prefix. There are a lot more with the `system-`prefix. Should I make that a separate type of retrieval?

Reply:
I'm not sure. The `system-illumos-` prefix was included because there are some of our services that still use the original naming of the SMF services, like `svc:/system/illumos/propolis-server`. (I think some of the networking daemons are still named this way too.)

Searching for that prefix becomes more reasonable when we're only looking for logs that come from a specific, named service, which is the case of the zone bundling process. That might be too permissive for `oxlog`, I'm really not sure. If that's true, then maybe we can commonize the portion of the checks that make sense (say, for SMF services named like `svc:/oxide/blah`). It seems like you might also be trying to capture _all_ log files for _any_ service, not just the "primary" Oxide services in zone. In that case, the commonization might be less helpful.

Reply:
> Searching for that prefix becomes more reasonable when we're only looking for logs that come from a specific, named service, which is the case of the zone bundling process. That might be too permissive for oxlog, I'm really not sure.

I think this is likely to be a common use case, but we also capture all logs in a zone. This means that we'll suck up all 'system-' prefixed logs unless we hardcode a list of services that can be named like this. I've so far avoiding hardcoding the names of any services and deducing service names by the names of the log-files. It turns out that strategy falls apart because of this dichotomy in naming. It seems to me that we would be better off changing the prefix for things like propolis to be `oxide-`, but in the meantime I can suck up only files that match a hardcoded set of services so we don't miss them. I specifically do not want to include all system service files as there are so many and mostly we won't want to dig through them in the common case. Again, though, we likely will want to be able to retrieve those files in some debugging scenarios though. Getting them through sled-agent integration via oxlog is preferable than having to ssh in. I will add some support for this as well.

>  It seems like you might also be trying to capture all log files for any service, not just the "primary" Oxide services in zone. In that case, the commonization might be less helpful.

I'm trying to gather any oxide specific service log in a zone, ideally without knowing the name of the service ahead of time. That may be too ambitious.

Reply:
> I'm trying to gather any oxide specific service log in a zone, ideally without knowing the name of the service ahead of time. That may be too ambitious.

So the way we did that in zone-bundling is to list all SMF FMRIs matching the patterns `svc:/oxide/*` or `svc:/system/illumos/*`, and then listing the logfiles associated with each of those (with `svcs -L`). 

Based on your answer, it does really sound like you want to use the same logic as in the zone bundling, or at least, use that logic possibly in addition to some other. Note that the prefix used in the zone-bundling is `system-illumos-*` _not_ `system-*`. The latter is definitely very permissive, and will get you a bunch of stuff unrelated to Oxide services. If the goal is Oxide-specific log files, then I think the logic can be shared with zone-bundling :)

Reply:
>  Note that the prefix used in the zone-bundling is system-illumos-* not system-*. The latter is definitely very permissive, and will get you a bunch of stuff unrelated to Oxide services. If the goal is Oxide-specific log files, then I think the logic can be shared with zone-bundling :)

Aha! Alright, so far the only dependency for oxlog is using the filesystem. But I can definitely parse using `system-illumos-*` and include those files. Then we can share the logic.

Reply:
I added this support in [c38d7e4](https://github.com/oxidecomputer/omicron/pull/4810/commits/c38d7e4c1f036546bf302d44e5c036d7fc278baf)

Reply:
`illumos-utils` now uses the method which I copied into oxlog from oxlog as of [caf3cd5](https://github.com/oxidecomputer/omicron/pull/4810/commits/caf3cd5f74be8313f00d617d9cdb4d08d11a9c41)